test/stdtypes/concurrency_test.go

```diff
@@ -0,0 +1,51 @@
+package stdtypes
+
+import (
+	"io/ioutil"
+	"sync"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+func TestConcurrentTextMarshal(t *testing.T) {
```

Comment:
The test requires -race to trigger the race condition.  Running this test against master yields:

```
# No race detector
$ go test ./test/stdtypes/
ok  	github.com/gogo/protobuf/test/stdtypes	0.032s

# With race detector
$ go test -race ./test/stdtypes/
==================
WARNING: DATA RACE
Read at 0x00c4200cc358 by goroutine 8:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:525 +0xc4d
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func2()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:33 +0xa7

Previous write at 0x00c4200cc358 by goroutine 7:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:534 +0xd3d
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func1()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:24 +0xa7

Goroutine 8 (running) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:37 +0x29c
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107

Goroutine 7 (running) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:28 +0x242
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107
==================
==================
WARNING: DATA RACE
Read at 0x00c4200cc4b9 by goroutine 8:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:538 +0xf66
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func2()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:33 +0xa7

Previous write at 0x00c4200cc4b9 by goroutine 7:
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeAny()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:546 +0x1104
  github.com/gogo/protobuf/proto.(*TextMarshaler).writeStruct()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:461 +0x13e1
  github.com/gogo/protobuf/proto.(*TextMarshaler).Marshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/proto/text.go:894 +0x4aa
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal.func1()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:24 +0xa7

Goroutine 8 (running) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:37 +0x29c
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107

Goroutine 7 (finished) created at:
  github.com/gogo/protobuf/test/stdtypes.TestConcurrentTextMarshal()
      /Users/snguyen/Sync/go/src/github.com/gogo/protobuf/test/stdtypes/concurrent_test.go:28 +0x242
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestConcurrentTextMarshal (0.00s)
	testing.go:610: race detected during execution of test
FAIL
FAIL	github.com/gogo/protobuf/test/stdtypes	0.061s
```